{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Mining Project - MS data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sean Deloddere - 914691 <br>\n",
    "Giulio Marcon - "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams[\"figure.figsize\"] = (20,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>f1</th>\n",
       "      <th>f2</th>\n",
       "      <th>f3</th>\n",
       "      <th>f4</th>\n",
       "      <th>f5</th>\n",
       "      <th>f6</th>\n",
       "      <th>f7</th>\n",
       "      <th>f8</th>\n",
       "      <th>f9</th>\n",
       "      <th>...</th>\n",
       "      <th>f4991</th>\n",
       "      <th>f4992</th>\n",
       "      <th>f4993</th>\n",
       "      <th>f4994</th>\n",
       "      <th>f4995</th>\n",
       "      <th>f4996</th>\n",
       "      <th>f4997</th>\n",
       "      <th>f4998</th>\n",
       "      <th>f4999</th>\n",
       "      <th>f5000</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>15.9163</td>\n",
       "      <td>14.9321</td>\n",
       "      <td>15.0390</td>\n",
       "      <td>17.8641</td>\n",
       "      <td>13.9613</td>\n",
       "      <td>13.7340</td>\n",
       "      <td>16.0623</td>\n",
       "      <td>18.8601</td>\n",
       "      <td>13.2050</td>\n",
       "      <td>...</td>\n",
       "      <td>14.6288</td>\n",
       "      <td>15.3938</td>\n",
       "      <td>16.5050</td>\n",
       "      <td>15.5140</td>\n",
       "      <td>15.3419</td>\n",
       "      <td>16.1521</td>\n",
       "      <td>13.5364</td>\n",
       "      <td>17.3797</td>\n",
       "      <td>11.0531</td>\n",
       "      <td>17.6632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16.7356</td>\n",
       "      <td>15.4755</td>\n",
       "      <td>16.3641</td>\n",
       "      <td>17.1888</td>\n",
       "      <td>15.6816</td>\n",
       "      <td>13.8570</td>\n",
       "      <td>15.3676</td>\n",
       "      <td>18.7065</td>\n",
       "      <td>13.1914</td>\n",
       "      <td>...</td>\n",
       "      <td>14.3400</td>\n",
       "      <td>16.9335</td>\n",
       "      <td>13.9464</td>\n",
       "      <td>15.4398</td>\n",
       "      <td>15.3828</td>\n",
       "      <td>14.6178</td>\n",
       "      <td>14.3196</td>\n",
       "      <td>19.1234</td>\n",
       "      <td>11.6904</td>\n",
       "      <td>18.9757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16.0818</td>\n",
       "      <td>15.5128</td>\n",
       "      <td>16.0463</td>\n",
       "      <td>17.7098</td>\n",
       "      <td>16.3100</td>\n",
       "      <td>14.0004</td>\n",
       "      <td>15.2811</td>\n",
       "      <td>18.2759</td>\n",
       "      <td>14.3015</td>\n",
       "      <td>...</td>\n",
       "      <td>14.4362</td>\n",
       "      <td>17.3463</td>\n",
       "      <td>13.8546</td>\n",
       "      <td>15.2012</td>\n",
       "      <td>15.1175</td>\n",
       "      <td>15.4171</td>\n",
       "      <td>14.4186</td>\n",
       "      <td>18.3941</td>\n",
       "      <td>11.5052</td>\n",
       "      <td>18.2264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16.5298</td>\n",
       "      <td>14.4695</td>\n",
       "      <td>15.0169</td>\n",
       "      <td>18.9366</td>\n",
       "      <td>17.2052</td>\n",
       "      <td>14.0647</td>\n",
       "      <td>16.7984</td>\n",
       "      <td>16.8349</td>\n",
       "      <td>13.2690</td>\n",
       "      <td>...</td>\n",
       "      <td>14.2350</td>\n",
       "      <td>15.6478</td>\n",
       "      <td>12.4729</td>\n",
       "      <td>16.1781</td>\n",
       "      <td>14.4310</td>\n",
       "      <td>15.7018</td>\n",
       "      <td>13.4911</td>\n",
       "      <td>13.3618</td>\n",
       "      <td>9.5018</td>\n",
       "      <td>17.4410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>16.6449</td>\n",
       "      <td>15.2317</td>\n",
       "      <td>16.1317</td>\n",
       "      <td>16.6674</td>\n",
       "      <td>14.0423</td>\n",
       "      <td>13.5085</td>\n",
       "      <td>15.2629</td>\n",
       "      <td>18.2195</td>\n",
       "      <td>12.3015</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0427</td>\n",
       "      <td>17.0514</td>\n",
       "      <td>14.9944</td>\n",
       "      <td>15.1685</td>\n",
       "      <td>15.0767</td>\n",
       "      <td>15.1256</td>\n",
       "      <td>14.0295</td>\n",
       "      <td>19.6029</td>\n",
       "      <td>9.1163</td>\n",
       "      <td>18.4286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 5001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    class       f1       f2       f3       f4       f5       f6       f7  \\\n",
       "id                                                                         \n",
       "1       1  15.9163  14.9321  15.0390  17.8641  13.9613  13.7340  16.0623   \n",
       "2       1  16.7356  15.4755  16.3641  17.1888  15.6816  13.8570  15.3676   \n",
       "3       1  16.0818  15.5128  16.0463  17.7098  16.3100  14.0004  15.2811   \n",
       "4       1  16.5298  14.4695  15.0169  18.9366  17.2052  14.0647  16.7984   \n",
       "5       1  16.6449  15.2317  16.1317  16.6674  14.0423  13.5085  15.2629   \n",
       "\n",
       "         f8       f9  ...    f4991    f4992    f4993    f4994    f4995  \\\n",
       "id                    ...                                                \n",
       "1   18.8601  13.2050  ...  14.6288  15.3938  16.5050  15.5140  15.3419   \n",
       "2   18.7065  13.1914  ...  14.3400  16.9335  13.9464  15.4398  15.3828   \n",
       "3   18.2759  14.3015  ...  14.4362  17.3463  13.8546  15.2012  15.1175   \n",
       "4   16.8349  13.2690  ...  14.2350  15.6478  12.4729  16.1781  14.4310   \n",
       "5   18.2195  12.3015  ...  14.0427  17.0514  14.9944  15.1685  15.0767   \n",
       "\n",
       "      f4996    f4997    f4998    f4999    f5000  \n",
       "id                                               \n",
       "1   16.1521  13.5364  17.3797  11.0531  17.6632  \n",
       "2   14.6178  14.3196  19.1234  11.6904  18.9757  \n",
       "3   15.4171  14.4186  18.3941  11.5052  18.2264  \n",
       "4   15.7018  13.4911  13.3618   9.5018  17.4410  \n",
       "5   15.1256  14.0295  19.6029   9.1163  18.4286  \n",
       "\n",
       "[5 rows x 5001 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ms_df = pd.read_csv('msdata.csv', index_col = 0)\n",
    "ms_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ms dataframe has 0 columns with missing values.\n"
     ]
    }
   ],
   "source": [
    "ms_cols_with_missing = [col for col in ms_df.columns if ms_df[col].isnull().any()]\n",
    "\n",
    "print(\"ms dataframe has \"+str(len(ms_cols_with_missing))+\" columns with missing values.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(ms_df.columns) == 5001:\n",
    "    true_labels = ms_df['class']\n",
    "    ms_df = ms_df.drop('class',axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[15.9163 14.9321 15.039  ... 17.3797 11.0531 17.6632]\n",
      " [16.7356 15.4755 16.3641 ... 19.1234 11.6904 18.9757]\n",
      " [16.0818 15.5128 16.0463 ... 18.3941 11.5052 18.2264]\n",
      " ...\n",
      " [16.8768 14.7009 16.0331 ... 18.0713  7.4234 17.4883]\n",
      " [16.6324  8.714  15.4801 ... 19.1017  5.4156 10.7977]\n",
      " [16.5564  9.7216 15.4208 ... 18.1179 10.9165 17.3568]]\n"
     ]
    }
   ],
   "source": [
    "ms = ms_df.to_numpy()\n",
    "ms = ms.astype(np.float)\n",
    "print(ms)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from sklearn.metrics.pairwise import euclidean_distances\n",
    "from sklearn.neighbors import DistanceMetric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "ss = StandardScaler()\n",
    "ms_ss = ss.fit_transform(ms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "mas = MaxAbsScaler()\n",
    "ms_mas = mas.fit_transform(ms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms_cs = cosine_similarity(ms)\n",
    "ms_cs_ss = cosine_similarity(ms_ss)\n",
    "ms_cs_mas = cosine_similarity(ms_mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def euclidean_similarity(data):\n",
    "    return 1/(1+euclidean_distances(data, data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms_ed = euclidean_similarity(ms)\n",
    "ms_ed_ss = euclidean_similarity(ms_ss)\n",
    "ms_ed_mas = euclidean_similarity(ms_mas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics.cluster import normalized_mutual_info_score\n",
    "from sklearn.cluster import SpectralClustering\n",
    "from sklearn.cluster import AgglomerativeClustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nmi(labels1,labels2):\n",
    "    return normalized_mutual_info_score(labels1,labels2,average_method='geometric')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### K-Means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_KMeans(data):\n",
    "    kmeans_list = []\n",
    "    for i in range(1,9):\n",
    "        kmeans = KMeans(n_clusters=i).fit(data)\n",
    "        labels = kmeans.labels_\n",
    "        kmeans_list.append((i,get_nmi(true_labels,labels),labels))\n",
    "    sorted_kmeans_list = sorted(kmeans_list, key=lambda x:x[1], reverse=True)\n",
    "    return sorted_kmeans_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmeans_0 = perform_KMeans(ms)\n",
    "kmeans_ss = perform_KMeans(ms_ss)\n",
    "kmeans_mas = perform_KMeans(ms_mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "(1, 0.25)\n",
      "(2, 0.45852623160823597)\n",
      "(2, 0.578193197941983)\n",
      "1\n",
      "(2, 0.12542231652480493)\n",
      "(3, 0.3916110216324213)\n",
      "(1, 0.25)\n",
      "2\n",
      "(7, 0.12175445919896208)\n",
      "(1, 0.25)\n",
      "(7, 0.20686722299017668)\n"
     ]
    }
   ],
   "source": [
    "for i in range(3):\n",
    "    print(i)\n",
    "    print(kmeans_0[i][0:2])\n",
    "    print(kmeans_ss[i][0:2])\n",
    "    print(kmeans_mas[i][0:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hierarchical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_hierarchical(data,linkage):\n",
    "    hier_list = []\n",
    "    for i in range(1,9):\n",
    "        hier = AgglomerativeClustering(n_clusters=i,linkage=linkage).fit(data)\n",
    "        labels = hier.labels_\n",
    "        hier_list.append((i,get_nmi(true_labels,labels),labels))\n",
    "    sorted_hier_list = sorted(hier_list, key=lambda x:x[1], reverse=True)\n",
    "    return sorted_hier_list        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ward\n",
      "complete\n",
      "average\n",
      "single\n"
     ]
    }
   ],
   "source": [
    "linkages = ['ward','complete','average','single']\n",
    "hier_0 = []\n",
    "hier_ss = []\n",
    "hier_mas = []\n",
    "for linkage in linkages:\n",
    "    print(linkage)\n",
    "    hier_0.append((linkage,perform_hierarchical(ms,linkage)))\n",
    "    hier_ss.append((linkage,perform_hierarchical(ms_ss,linkage)))\n",
    "    hier_mas.append((linkage,perform_hierarchical(ms_mas,linkage)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================\n",
      "ward\n",
      "0\n",
      "(5, 0.4739938347242577)\n",
      "(6, 0.44632441019560576)\n",
      "(7, 0.4277016120569969)\n",
      "ss\n",
      "(6, 0.3840490366198332)\n",
      "(7, 0.3695913426247361)\n",
      "(8, 0.360991476291801)\n",
      "mas\n",
      "(4, 0.5826743412077233)\n",
      "(5, 0.5444413723904372)\n",
      "(6, 0.5052375371794963)\n",
      "============================================\n",
      "complete\n",
      "0\n",
      "(8, 0.2683424053873574)\n",
      "(1, 0.25)\n",
      "(5, 0.24801554690137462)\n",
      "ss\n",
      "(7, 0.255719325551191)\n",
      "(8, 0.2507309209089276)\n",
      "(1, 0.25)\n",
      "mas\n",
      "(1, 0.25)\n",
      "(7, 0.24785083275381253)\n",
      "(6, 0.24457616340611107)\n",
      "============================================\n",
      "average\n",
      "0\n",
      "(1, 0.25)\n",
      "(8, 0.03572245741022434)\n",
      "(7, 0.03302978442800275)\n",
      "ss\n",
      "(1, 0.25)\n",
      "(8, 0.03572245741022434)\n",
      "(7, 0.03302978442800275)\n",
      "mas\n",
      "(1, 0.25)\n",
      "(8, 0.03572245741022434)\n",
      "(7, 0.03302978442800275)\n",
      "============================================\n",
      "single\n",
      "0\n",
      "(1, 0.25)\n",
      "(8, 0.03572245741022434)\n",
      "(7, 0.03302978442800275)\n",
      "ss\n",
      "(1, 0.25)\n",
      "(8, 0.03467767612162472)\n",
      "(7, 0.0319232585145711)\n",
      "mas\n",
      "(1, 0.25)\n",
      "(8, 0.03572245741022434)\n",
      "(7, 0.03302978442800275)\n"
     ]
    }
   ],
   "source": [
    "for i in range(4):\n",
    "    print(\"============================================\")\n",
    "    print(linkages[i])\n",
    "    print(\"0\")\n",
    "    for j in range(3):\n",
    "        print(hier_0[i][1][j][0:2])\n",
    "    print(\"ss\")\n",
    "    for j in range(3):\n",
    "        print(hier_ss[i][1][j][0:2])\n",
    "    print(\"mas\")\n",
    "    for j in range(3):\n",
    "        print(hier_mas[i][1][j][0:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spectral "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Nearest Neighbor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_spectral_nn(data):\n",
    "    spectral_list = []\n",
    "    for nc in range(2,8):\n",
    "        for nn in range(5,9,2):\n",
    "            spectral = SpectralClustering(n_clusters=nc, affinity='nearest_neighbors',n_neighbors=nn,n_jobs=3).fit(data)\n",
    "            labels = spectral.labels_\n",
    "            spectral_list.append((nc,nn,get_nmi(true_labels,labels),labels))\n",
    "    sorted_spectral_list = sorted(spectral_list, key=lambda x:x[2], reverse=True)\n",
    "    return sorted_spectral_list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n"
     ]
    }
   ],
   "source": [
    "spectral_0 = perform_spectral_nn(ms)\n",
    "spectral_ss = perform_spectral_nn(ms_ss)\n",
    "spectral_mas = perform_spectral_nn(ms_mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "(4, 5, 0.042587356028537034)\n",
      "(4, 7, 0.042587356028537034)\n",
      "(6, 7, 0.022241101442191203)\n",
      "ss\n",
      "(4, 5, 0.042587356028537034)\n",
      "(4, 7, 0.028135934702172953)\n",
      "(5, 5, 0.020422713584982714)\n",
      "mas\n",
      "(4, 5, 0.042587356028537034)\n",
      "(4, 7, 0.03238275974538477)\n",
      "(5, 5, 0.020517634177940405)\n"
     ]
    }
   ],
   "source": [
    "print(\"0\")\n",
    "for i in range(3):\n",
    "    print(spectral_0[i][0:3])\n",
    "\n",
    "print(\"ss\")\n",
    "for i in range(3):\n",
    "    print(spectral_ss[i][0:3])\n",
    "\n",
    "print(\"mas\")\n",
    "for i in range(3):\n",
    "    print(spectral_mas[i][0:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### RBF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_spectral_rbf(data):\n",
    "    spectral_list = []\n",
    "    for nc in range(2,8):\n",
    "        spectral = SpectralClustering(n_clusters=nc, affinity='rbf',n_jobs=3).fit(data)\n",
    "        labels = spectral.labels_\n",
    "        spectral_list.append((nc,get_nmi(true_labels,labels),labels))\n",
    "    sorted_spectral_list = sorted(spectral_list, key=lambda x:x[1], reverse=True)\n",
    "    return sorted_spectral_list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/spectral.py:268: ConvergenceWarning: Number of distinct clusters (2) found smaller than n_clusters (4). Possibly due to duplicate points in X.\n",
      "  n_init=n_init)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/numpy/core/_methods.py:121: RuntimeWarning: overflow encountered in multiply\n",
      "  x = um.multiply(x, x, out=x)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/spectral.py:268: ConvergenceWarning: Number of distinct clusters (1) found smaller than n_clusters (5). Possibly due to duplicate points in X.\n",
      "  n_init=n_init)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/numpy/core/_methods.py:121: RuntimeWarning: overflow encountered in multiply\n",
      "  x = um.multiply(x, x, out=x)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:273: RuntimeWarning: overflow encountered in multiply\n",
      "  distances = - 2 * safe_sparse_dot(X, Y.T, dense_output=True)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/spectral.py:268: ConvergenceWarning: Number of distinct clusters (5) found smaller than n_clusters (6). Possibly due to duplicate points in X.\n",
      "  n_init=n_init)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/manifold/spectral_embedding_.py:235: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.\n",
      "  warnings.warn(\"Graph is not fully connected, spectral embedding\"\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/numpy/core/_methods.py:121: RuntimeWarning: overflow encountered in multiply\n",
      "  x = um.multiply(x, x, out=x)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: overflow encountered in square\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:442: RuntimeWarning: invalid value encountered in subtract\n",
      "  max_iter=max_iter, verbose=verbose)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/metrics/pairwise.py:274: RuntimeWarning: invalid value encountered in add\n",
      "  distances += XX\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/k_means_.py:444: RuntimeWarning: overflow encountered in square\n",
      "  inertia = np.sum((X - centers[labels]) ** 2, dtype=np.float64)\n",
      "/Users/giuliomarcon/anaconda3/lib/python3.7/site-packages/sklearn/cluster/spectral.py:268: ConvergenceWarning: Number of distinct clusters (1) found smaller than n_clusters (7). Possibly due to duplicate points in X.\n",
      "  n_init=n_init)\n"
     ]
    }
   ],
   "source": [
    "spectral_rbf_0 = perform_spectral_rbf(ms)\n",
    "spectral_rbf_ss = perform_spectral_rbf(ms_ss)\n",
    "spectral_rbf_mas = perform_spectral_rbf(ms_mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "(7, 0.006707050104335941)\n",
      "(4, 0.004726163376626843)\n",
      "(6, 0.0028971802448228244)\n",
      "ss\n",
      "(5, 0.25)\n",
      "(7, 0.25)\n",
      "(6, 0.026899214419641385)\n",
      "mas\n",
      "(7, 0.08821550650906979)\n",
      "(6, 0.05773301346740331)\n",
      "(5, 0.05700304664619385)\n"
     ]
    }
   ],
   "source": [
    "print(\"0\")\n",
    "for i in range(3):\n",
    "    print(spectral_rbf_0[i][0:2])\n",
    "\n",
    "print(\"ss\")\n",
    "for i in range(3):\n",
    "    print(spectral_rbf_ss[i][0:2])\n",
    "\n",
    "print(\"mas\") \n",
    "for i in range(3):\n",
    "    print(spectral_rbf_mas[i][0:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Precomputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_spectral_pc(data):\n",
    "    spectral_list = []\n",
    "    for nc in range(2,8):\n",
    "        spectral = SpectralClustering(n_clusters=nc, affinity='precomputed',n_jobs=3).fit(data)\n",
    "        labels = spectral.labels_\n",
    "        spectral_list.append((nc,get_nmi(true_labels,labels),labels))\n",
    "    sorted_spectral_list = sorted(spectral_list, key=lambda x:x[1], reverse=True)\n",
    "    return sorted_spectral_list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "spectral_pc_cs = perform_spectral_pc(ms_cs)\n",
    "#spectral_pc_cs_ss = perform_spectral_pc(ms_cs_ss)\n",
    "spectral_pc_cs_mas = perform_spectral_pc(ms_cs_mas)\n",
    "spectral_pc_ed = perform_spectral_pc(ms_ed)\n",
    "spectral_pc_ed_ss = perform_spectral_pc(ms_ed_ss)\n",
    "spectral_pc_ed_mas = perform_spectral_pc(ms_ed_mas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cs 0\n",
      "(3, 0.8477183741661061)\n",
      "(3, 0.8477183741661061)\n",
      "(3, 0.8477183741661061)\n",
      "cs mas\n",
      "(3, 0.8748134474580916)\n",
      "(3, 0.8748134474580916)\n",
      "(3, 0.8748134474580916)\n",
      "ed 0\n",
      "(6, 0.8802085025252647)\n",
      "(6, 0.8802085025252647)\n",
      "(6, 0.8802085025252647)\n",
      "ed ss\n",
      "(6, 0.9085154581352097)\n",
      "(6, 0.9085154581352097)\n",
      "(6, 0.9085154581352097)\n",
      "ed mas\n",
      "(5, 0.9150261357661318)\n",
      "(5, 0.9150261357661318)\n",
      "(5, 0.9150261357661318)\n"
     ]
    }
   ],
   "source": [
    "print(\"cs 0\")\n",
    "for i in range(3):\n",
    "    print(spectral_pc_cs[0][0:2])\n",
    "#print(spectral_pc_cs_ss[0][0:2])\n",
    "print(\"cs mas\")\n",
    "for i in range(3):\n",
    "    print(spectral_pc_cs_mas[0][0:2])\n",
    "print(\"ed 0\")\n",
    "for i in range(3):\n",
    "    print(spectral_pc_ed[0][0:2])\n",
    "print(\"ed ss\")\n",
    "for i in range(3):\n",
    "    print(spectral_pc_ed_ss[0][0:2])\n",
    "print(\"ed mas\")\n",
    "for i in range(3):\n",
    "    print(spectral_pc_ed_mas[0][0:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BEST METHOD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MaxAbsScaler + Euclidean Similarity + Spectral Clustering (5 clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3 3 3 3 3 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1\n",
      " 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1\n",
      " 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1\n",
      " 1 1 1 1 1 1 1 1 1 1 1 1 1 1 4 3 4 3 4 2 4 3 3 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2\n",
      " 2 2 2 2 2 2 2 2 2 2 2 2 2 2 4 4 4 4 4 4 3 4 3 3 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "Best_labels = spectral_pc_ed_mas[0][2]\n",
    "print(Best_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('ms_solution.txt','w') as file:\n",
    "    for label in Best_labels:\n",
    "        file.write(str(label)+'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
